test/tyxml_test/element_test.ml

```diff
@@ -0,0 +1,335 @@
+    let expected =
+      iframe
+        ~a:[ a_src "https://github.com/xvw/preface"; a_name "frame-2" ]
+        [ txt "test" ]
+    and computed =
+      iframe_of ~a:[ a_name "frame-2" ] E.repo "xvw" "preface" [ txt "test" ]
+    in
+    expected, computed)
+;;
+
+let test_img_of =
+  test_equality
+    ~about:"img_of"
+    ~desc:"check that generated HTML has the right form"
+    html_testable
+    (fun () ->
+    let open Tyxml.Html in
+    let open Nightmare_tyxml in
+    let expected = img ~alt:"a kip" ~src:"/files/media/kip.jpg" ()
+    and computed = img_of E.media "media" "kip.jpg" ~alt:"a kip" in
+    expected, computed)
+;;
+
+let test_link_of =
+  test_equality
+    ~about:"link_of"
+    ~desc:"check that generated HTML has the right form"
+    html_testable
+    (fun () ->
+    let open Tyxml.Html in
+    let open Nightmare_tyxml in
+    let expected = link ~rel:[ `Stylesheet ] ~href:"/files/css/default.css" ()
+    and computed = link_of ~rel:[ `Stylesheet ] E.media "css" "default.css" in
+    expected, computed)
+;;
+
+let test_object_of =
+  test_equality
+    ~about:"object_of"
+    ~desc:"check that generated HTML has the right form"
+    html_testable
+    (fun () ->
+    let open Tyxml.Html in
+    let open Nightmare_tyxml in
+    let expected =
+      object_
+        ~a:
+          [ a_data "/files/media/clip.mp4"
+          ; a_width 200
+          ; a_height 300
+          ; a_mime_type "video/mp4"
+          ]
+        [ txt "foo" ]
+    and computed =
+      object_of
+        ~a:[ a_width 200; a_height 300; a_mime_type "video/mp4" ]
+        E.media
+        "media"
+        "clip.mp4"
+        [ txt "foo" ]
+    in
+    expected, computed)
+;;
+
+let test_script_of =
+  test_equality
+    ~about:"script_of"
+    ~desc:"check that generated HTML has the right form"
+    html_testable
+    (fun () ->
+    let open Tyxml.Html in
+    let open Nightmare_tyxml in
+    let expected =
+      script
+        ~a:[ a_src "/files/js/front.js"; a_mime_type "application/javascript" ]
+        (txt "hello")
+    and computed =
+      script_of
+        ~a:[ a_mime_type "application/javascript" ]
+        E.media
+        "js"
+        "front.js"
+        (txt "hello")
+    in
+    expected, computed)
+;;
+
+let test_source_of =
+  test_equality
+    ~about:"source_of"
+    ~desc:"check that generated HTML has the right form"
+    html_testable
+    (fun () ->
+    let open Tyxml.Html in
+    let open Nightmare_tyxml in
+    let expected =
+      source ~a:[ a_src "/files/video/movie.mp4"; a_mime_type "video/mp4" ] ()
+    and computed =
+      source_of ~a:[ a_mime_type "video/mp4" ] E.media "video" "movie.mp4"
+    in
+    expected, computed)
+;;
+
+let cases =
+  ( "Element"
+  , [ test_a_of
+    ; test_base_of
+    ; test_embed_of
+    ; test_form_of
+    ; test_form_with_csrf_of
+    ; test_iframe_of
+    ; test_img_of
+    ; test_link_of
+    ; test_object_of
+    ; test_script_of
+    ; test_source_of
+    ] )
+;;
```

Comment:
Not really related but I think that we should keep in mind that it could be useful to have cram-like test for, for example, testing polymorphic variant exclusion. (But you can resolve this thread)

Reply:
Yes, after `form`, I'll take some times to think about testing strategy, benchmarks and CI.